Allow round trip of Time, TimeDelta and SkyCoord through ECSV file

taldcroft opened the pull request:
This PR adds a new `astropy-2.0` schema to the ECSV format to allow (mostly) lossless serialization of `Time`, `TimeDelta`, and `SkyCoord` objects.  The "mostly" means that the representation of numbers if just what the user specifies, so if a Time object has precision=1 then that's what get's written out.

For example of this in action see below.

To do:

- [x] Set `schema` keyword in ECSV header to `astropy-2.0`.
- [x] Tests
- [x] Docs
- [x] Changelog

```
In [27]: tm = Time([1,2], format='cxcsec', precision=5)
In [28]: sc = SkyCoord([1,2], [3,4], unit='deg,deg', frame='fk4', obstime='J1990.5')
In [29]: t = Table([sc, tm])
In [30]: t.write('mixin.ecsv', format='ascii.ecsv', overwrite=True)
In [31]: cat mixin.ecsv
# %ECSV 0.9
# ---
# datatype:
# - name: col0
#   unit: deg,deg
#   datatype: object
#   meta:
#     __object_attributes__:
#       class: astropy.coordinates.sky_coordinate.SkyCoord
#       data_attrs: [ra, dec]
#       datatype: string
#       equinox: !astropy.time.Time {format: byear_str, in_subfmt: '*', jd1: 2400000.5,
#         jd2: 33281.92345905, out_subfmt: '*', precision: 3, scale: tai}
#       frame: fk4
#       obstime: !astropy.time.Time {format: jyear_str, in_subfmt: '*', jd1: 2400000.5,
#         jd2: 48074.625, out_subfmt: '*', precision: 3, scale: utc}
#       representation: spherical
#       unit: deg,deg
# - name: col1
#   datatype: object
#   meta:
#     __object_attributes__:
#       class: astropy.time.core.Time
#       data_attrs: [val]
#       datatype: float64
#       format: cxcsec
#       in_subfmt: '*'
#       out_subfmt: '*'
#       precision: 5
#       scale: tt
col0 col1
1.0,3.0 1.000000000001755
2.0,4.0 1.999999999998714

In [32]: t2 = Table.read('mixin.ecsv', format='ascii.ecsv')

In [33]: t2
Out[33]: 
<Table length=2>
  col0         col1      
deg,deg                  
 object       object     
------- -----------------
1.0,3.0 1.000000000001755
2.0,4.0 1.999999999998714

In [34]: t2['col0']
Out[34]: 
<SkyCoord (FK4: equinox=B1950.000, obstime=J1990.500): (ra, dec) in deg
    [( 1.,  3.), ( 2.,  4.)]>

In [35]: t2['col1']
Out[35]: 
<Time object: scale='tt' format='cxcsec' value=col1
----
 1.0
 2.0>
```

mhvk commented:
@taldcroft - looks good overall. I'm still at a meeting so didn't have time to try, so just some questions about functionality:

- What happens for multidimensional coordinate attributes? As currently written, attributes can either have no shape, or the same shape as the coordinate, so one option would be to write the attributes together with the coordinates proper.
- Even for single-element attributes, should, e.g., the `Time` attributes be formatted similarly to what happens to `Time` columns?
- I guess the same holds for `Time` locations.
- I like that for Time the default format is what was used as input. But is it *possible* to get the jd1,jd2 format? Or is the only way to keep precision using isot with precision=9?
- For coordinates, it would be nice if one could similarly control how it looks. Am I right in guessing that currently it is not possible to have the coordinates as hms, dms pairs? (I did a quick test but the `hourangle` units I put in seem to get converted to degrees.)

p.s. I think the last two items are probably out of scope for this PR: we need some way to give use formats in `SkyCoord`, and for `Time`, the easiest route may well be to create a `TimeTwoPartJD` format.

taldcroft commented:
> What happens for multidimensional coordinate attributes?

Attributes are just stored as pure YAML objects regardless of dimensionality.  So indeed if you have a million different `obstime` values this going to result in a long and ugly header.  It is not immediately clear to me if this will actually *break*.  A test with 10000 was fine, but going to 100000 took over a minute of crunching before I killed it.   But for now in the 90% use case this implementation will work reliably.

> Even for single-element attributes, should, e.g., the Time attributes be formatted similarly to what happens to Time columns?

Are you talking about something like `equinox` in the example above?  The goal is to do the whole header in a straight-YAML representation to make decoding as simple as possible.  (Not only for astropy but for non-astropy readers).

> I like that for Time the default format is what was used as input. But is it possible to get the jd1,jd2 format? Or is the only way to keep precision using isot with precision=9?

Not at the moment, but I agree on the need for this.  See below.

> For coordinates, it would be nice if one could similarly control how it looks. Am I right in guessing that currently it is not possible to have the coordinates as hms, dms pairs? (I did a quick test but the hourangle units I put in seem to get converted to degrees.)

Probably not until coordinates / angles get an intrinsic `format` property.  I'm 100% behind this idea and argued strongly for that years ago, but it never got traction (and I never just did it myself).

BTW, the way that I really want to do this is to allow for the concept of a multi-column which is an object that gets broken into multiple separate columns for writing and then constructed from those columns for reading.  So my example above would have data like:
```
col0.ra col0.dec col1
1.0 3.0 1.000000000001755
2.0 4.0 1.999999999998714
```
I think this is not *that* difficult, but not trivial either.  This could solve the problem of attributes that have the same length as the column, but you do then have to deal with nested structures.  For example a `location` has three attributes, so it would be:
```
col0.ra col0.dec col0.location.x col0.location.y col0.location.z col1
1.0 3.0 1 2 3 1.000000000001755
2.0 4.0 3 4 5 1.999999999998714
```
I do have some work done on those lines but this is definitely not for 2.0.

taldcroft commented:
I may have an idea for better implementation..

mhvk commented:
I'll await the different idea, but I like the idea of column names with parts; one can then think of these as parts that help reconstruct the object, as in
```
reconstructed = m_oa_['class'](jd1=col0.jd1, jd2=col0.jd2, format=m_oa['format'],
                               location=m_oa['location']['class'](col0.location.x, col0.location.y, col0.location.z)
```
(`m_oa` is `meta['__object_attributes__']` -- obviously not correct in detail, but I like the correspondence with reconstruction)

taldcroft commented:
@mhvk - I finally have something to look at.  I know you are swamped but when you have a few minutes perhaps just thoughts about the overall concept.  Basically one represents a table with mixins via a table with plain columns + appropriate metadata.  That representation (+meta) gets written to the ECSV file.  The metadata is sufficient to then construct the original mixin-table from the plain-column version.

I think it's *much* better now, and the main action is now localized in:
- Info classes to define data and object attributes
- New `table.serialize` module that does the conversion to/from the intermediate representation.

So this is defining a new ECSV `schema` which I propose calling `astropy-2.0`.  This is an allowed keyword in ECSV and I will define it (but it's not yet done).  The schema is a particular way of specifying the attributes and data to construct a mixin column within the table `meta`.  All this is "hidden" in a key `__mixin_columns__` which is written that way to avoid stomping on user meta data (since it shares the same space with the user meta). 

So a key question is if this schema is OK to commit to.  Of course we can always refine and have `astropy-3.0` schema, but it would be better to avoid that.

taldcroft commented:
A notebook that demonstrates this is available here:

  https://github.com/taldcroft/astropy/blob/time-ecsv/mixin-ecsv.ipynb

I committed this to the branch here for convenience but will rebase eventually and remove it.

taldcroft commented:
@mhvk - you probably should not spend too much time looking at the `io.ascii` updates.  This package is a bit intricate but I think the way I've done it is fine and relatively clean.  I would like to go back and work on all this to make it less confusing, but not for now.

taldcroft commented:
In `serialize.py` there is a key thing to review in `_construct_mixin_from_obj_attrs(obj_attrs)`.  Namely the method that I use to avoid executing arbitrary code via a registry of allowed fully-qualified class names.  OK?

mhvk commented:
@taldcroft - I've been reviewing the proper motion PRs the whole evening here (in Stockholm), but hope to have some time tomorrow to look at this. 

Aside: The idea rather reminds me of the alternative Wolfgang and I had for implementing mixin columns (#1835), where we serialized all the arrays that described an astropy object as a structured array that was made into a `Column` - for `Table`, your solution was far superior, since you realized that tables can just hold things that have roughly columnlike properties, but here the structured array would seem to make a lot of sense, since an ascii file cannot possibly hold `Time` instances, but should be quite OK with "structured arrays". I look forward to checking it in more detail!

mhvk commented:
This looks very nice indeed! First, a high-level comment, which may influence thoughts on how to serialize array-shaped attributes:

I think it makes sense to think of the `__mixin_columns__` not so much as description of the *attributes* of what went in, but rather as a description of the *arguments* that would be needed to recreate the object (including some "hidden" arguments such as `_delta_ut1_utc`). Given that, two possibilities:
1. Reverse the order of the map linking to columns, ie., instead of `__data_attrs_map__`, have a `__kwarg_map__`, which contains for each argument what column it should get the values from.
2. A bit more drastic, but I think even better: do away with the mapping and instead, in the list of arguments, introduce a special way to link to a serialized column. This could be a string starting with some name, say `ra: __col__sc.ra`, or, guaranteed to not cause name conflicts and possibly also more logical in the context of how yaml serializes (if much more wordy), something like `!astropy.table.SerializedColumn {name: sc.ra}`.

With this, I think it will become easier to also serialize array-shaped attributes that themselves would be mixin columns, like `time.location` or `skycoord.obstime`. Trying to roughly think that through:
1. To recognize array-shaped attributes in the first place, it may be easiest to keep the full list of `_represent_as_dict_attrs` (hey, not even a hidden API change! Good in case anyone has made their own `info`) and select what gets put out as a full column in the serialized table by having a non-empty shape (here it helps that even for a one-row table, the shape of the *data* has to be `(1,)`, i.e., one can just get the shape of the first attribute in `_represent_as_dict_attrs` and add a column in the serialized table anything for which `first_shape == getattr(colattr, 'shape', ())`.
2. To represent mixin-like columns, extending the above, one might get (removing some simple attributes for brevity):
```
obstime: !astropy.time.Time {format: jyear_str, 
                             jd1: !astropy.table.SerializedColumn {name: sc.obstime.jd1},
                             jd2: !astropy.table.SerializedColumn {name: sc.obstime.jd2},
                             precision: 3, scale: utc}
```

taldcroft commented:
> I think it makes sense to think of the __mixin_columns__ not so much as description of the attributes of what went in, but rather as a description of the arguments that would be needed to recreate the object (including some "hidden" arguments such as _delta_ut1_utc). 

Agreed that there are two ways of thinking about serializing an object, either capturing the relevant object attributes, or capturing the required kwargs for initializing using the astropy constructor.  This dichotomy has been a source of some pain because they are not the same.  This requires class-specific code for the constructor in a number of cases (the `Info._construct_from_dict` method).

However, I will push back and say that I believe that serializing the object attributes is a conceptually more general and understandable way to do this.  The example of `Time` is a pretty good one, because if you just look at the output I think it is more useful to have `jd1` and `jd2` attributes than `val` and `val2` (which are the corresponding constructors).  In other words, one should focus on the *properties of the object*, rather than the details of the *astropy object constructor*.

You touched on this, but `EarthLocation` is example where you  simply cannot currently create the full object via `**kwargs` because of the `ellipsoid` attribute.  `Time` is another example where the constructor is insufficient to create the object.  I think that the current solution of handling this with custom code works reasonably well and hides the deficiencies of the astropy constructors from the ECSV / YAML output.

As a footnote the `_represent_as_dict_info_attrs` attribute is poorly named and should be `_represent_as_dict_object_attrs` as it really refers to things that live on the object (e.g. `EarthLocation.ellipsoid` rather than `Info` attributes like format.

taldcroft commented:
BTW, I'm still mulling over your bigger suggestion...

taldcroft commented:
On the idea of the `SerializedColumn` class, that seems nice and should work.  In the notebook example it would lead to this.
```
#     tm:
#       __class__: astropy.time.core.Time
#       __info__: {description: description-tm}
#       format: cxcsec
#       in_subfmt: '*'
#       jd1: !astropy.table.SerializedColumn {name: tm.jd1}
#       jd2: !astropy.table.SerializedColumn {name: tm.jd2}
#       out_subfmt: '*'
#       precision: 5
#       scale: tt
```
Just to be clear this has no impact at all on the existing pure-YAML serialization of core objects, but only affects the ECSV representation.  The latter is where we can put parts of the object into columnar format.

taldcroft commented:
For the 2.0 release the key is to get the API in a form that will work for subsequent development.  For instance I would punt on actually serializing array-like attributes (e.g. `sc.obstime`), but make sure header YAML format can handle it OK.  I think that is the case, and in the future constructing from either pure-YAML for (e.g.) `obstime` or from a column-derived version will both be supported.  How `obstime` gets serialized will depend on the shape.

mhvk commented:
@taldcroft - hmm, I think you are right that thinking of representing properties makes, in the end, more sense. It is perhaps similar to it not always being so easy to make `repr` do something that would actually reproduce the instance. So, let's stick with your approach.

On the `SerializedColumn` class, I do like the readability of the example you list above, that `jd1` and `jd2` are not in this `__data_attrs_map__`, but just part of the attributes of the time instance. I think it also makes extension to `location` easier: the "data" are now not so special any more.

It also seems that it could easily be used to store time in different formats, with something like
```
format : cxcsec
cxcsec : !astropy.table.SerializedColumn {name: tm.cxcsec}
```
(Not to do now, but just as an example that, I think, shows that this might be possible -- possibly `value` is more logical. In any case, it does seem another argument in favour of listing "attributes".)

Do you agree that with a scheme like the above, it should be possible, in principle, to store things like array-shaped `obstime` (e.g., something similarly to my example)?

mhvk commented:
p.s. Tiny thing, but it would be nice not to have `in_subfmt` and `out_subfmt` in the list if they are just `*`.

taldcroft commented:
My plan for `Time` was to stick with the `value` attribute, so just
```
format : cxcsec
value : !astropy.table.SerializedColumn {name: tm}
```
BTW, notice this special case and tell me if you agree.  Right now when there is exactly one data attribute (aka SerializedColumn) then it does NOT get the attribute name appended.  My thinking is that (esp. in the case of Time and Quantity) there is no need to change the column name.  This would be less confusing to users and prevents every Quantity column name from getting an extra `.value`.

Overall I think the SerializedColumn scheme will work (and I like it much better than the current scheme), but I guess I should try the implementation first!  I *think* it will be quite simple.

There is a little complication when getting to supporting array-shaped attributes like `sc.obstime` in the column data.  That is a `Time` object and has its own set of attributes.  So at this point one needs to reconstruct `sc` recursively.  I.e. the `sc` (skycoord) object has an `obstime` attribute which will be named `sc.obstime`, but that will only exist as an entry in the `__mixin_columns__`, where it will reference `sc.obstime.jd1` and `sc.obstime.jd2` (as well as including its own attributes like `frame` and `scale`).

I don't think this is a real problem, and I have been planning on recursive support like this (e.g. tables within tables), but it will be a bit more complicated.

mhvk commented:
When typing my example for `cxcsec`, I changed back and forth between that and `value` (since `repr` uses `value`, so yes, definitely fine to do that. I also agree with keeping a simple single name for the column if possible.

And recursion will be needed, but I also think that it indeed should not pose a real problem; certainly, in constructing the serialized table, I can see how one just recurses until all mixin columns have gone. And if that's possible, the reverse should too. Possibly, it might use another new type, say,  `astroy.table.MixinColumn`, with which the meta data might look like:
```
#     tm:
#       __class__: astropy.time.core.Time
#       __info__: {description: description-tm}
#       format: cxcsec
#       jd1: !astropy.table.SerializedColumn {name: tm.jd1}
#       jd2: !astropy.table.SerializedColumn {name: tm.jd2}
#       location: !astropy.table.MixinColumn {name: tm.location}
#       precision: 5
#       scale: tt
#    tm.location:
#       __class__: astropy.coordinates.earth.EarthLocation
#       ellipsoid: WGS84
#       x: !astropy.table.SerializedColumn {name: tm.location.x}
#       y: !astropy.table.SerializedColumn {name: tm.location.y}
#       z: !astropy.table.SerializedColumn {name: tm.location.z}
```
Not sure the `MixinColumn` class really is required, but it does make obvious that there is another level of recursion. In any case, most importantly, I cannot see how this would not work within the scheme, so should be fine to proceed!

taldcroft commented:
@mhvk - done with SerializedColumn and passing tests locally.
```
# %ECSV 0.9
# ---
# datatype:
# - {name: c, datatype: float64, format: .2f, description: description-c}
# - {name: tm.jd1, datatype: float64}
# - {name: tm.jd2, datatype: float64}
# - {name: sc.x, unit: '', datatype: float64}
# - {name: sc.y, unit: '', datatype: float64}
# - {name: sc.z, unit: '', datatype: float64}
# - {name: q, datatype: float64}
# - {name: lat, datatype: float64}
# - {name: ang, datatype: float64}
# - {name: el.x, unit: m, datatype: float64, description: description-el}
# - {name: el.y, unit: m, datatype: float64, description: description-el}
# - {name: el.z, unit: m, datatype: float64, description: description-el}
# meta: !!omap
# - __mixin_columns__:
#     ang:
#       __class__: astropy.coordinates.angles.Angle
#       __info__: {description: description-ang}
#       unit: &id001 !astropy.units.Unit {unit: deg}
#       value: !astropy.table.SerializedColumn {name: ang}
#     el:
#       __class__: astropy.coordinates.earth.EarthLocation
#       __info__: {description: description-el}
#       ellipsoid: WGS84
#       x: !astropy.table.SerializedColumn {name: el.x}
#       y: !astropy.table.SerializedColumn {name: el.y}
#       z: !astropy.table.SerializedColumn {name: el.z}
#     lat:
#       __class__: astropy.coordinates.angles.Latitude
#       __info__: {description: description-lat}
#       unit: *id001
#       value: !astropy.table.SerializedColumn {name: lat}
#     q:
#       __class__: astropy.units.quantity.Quantity
#       __info__: {description: description-q, format: .3f}
#       unit: !astropy.units.Unit {unit: m}
#       value: !astropy.table.SerializedColumn {name: q}
#     sc:
#       __class__: astropy.coordinates.sky_coordinate.SkyCoord
#       __info__: {description: description-sc}
#       frame: icrs
#       representation: cartesian
#       x: !astropy.table.SerializedColumn {name: sc.x}
#       y: !astropy.table.SerializedColumn {name: sc.y}
#       z: !astropy.table.SerializedColumn {name: sc.z}
#     tm:
#       __class__: astropy.time.core.Time
#       __info__: {description: description-tm}
#       format: cxcsec
#       in_subfmt: '*'
#       jd1: !astropy.table.SerializedColumn {name: tm.jd1}
#       jd2: !astropy.table.SerializedColumn {name: tm.jd2}
#       out_subfmt: '*'
#       precision: 5
#       scale: tt
c tm.jd1 tm.jd2 sc.x sc.y sc.z q lat ang el.x el.y el.z
1.00 2450815.0 -0.499988425926 0.998477438639 0.0174284885208 0.0523359562429 1.0 1.0 1.0 6368484.2824 111162.306644 331574.315343
2.00 2450815.0 -0.499976851852 0.998477438639 0.0174284885208 0.0523359562429 2.0 2.0 2.0 6358827.82377 222055.160671 441945.105248
```

taldcroft commented:
Right now having SerializedColumn be a dict (instead of just a scalar string) is overkill since `name` is the only key.  But maybe best play it safe and leave that flexibility?

mhvk commented:
This mostly looks great. One comment looking at your example: shouldn't we just put the unit of any quantity in the column? This happens for `el.[xyz]`, but somehow not for `q`, `ang`, and `lat`. Indeed, similarly for `format` and `description` -- essentially for any quantity subclass, all that should be needed is the class and where to find it -- the initializer can get everything from the Column. Maybe for description and other standard column information, this can even be generalized to any mixin for which only a single data column is needed?

Also, maybe good to add `Longitude` - it also needs `wrap_angle`.

taldcroft commented:
I addressed your comment about `unit`, which was a very good one because I realize this was changing the primary header info for Quantity in a bad way, i.e. losing the `unit` and other info.  It is obviously round-tripping fine but other non-astropy readers (or humans) would be lost.

I'm not 100% happy with the internal implementation, and we do end up repeating the info in the primary header and the `__mixin_columns__` info.  I think that fixing this is going to be tricky (see long code comment in the last commit).  In the end for 3.0 maybe we can improve the basic ECSV spec (ECSV-1.0!) to handle mixins and/or general aggregate columns.  What I would like is a new column `datatype` that specifies a more complex column.  But this is not for 2.0.

```
# %ECSV 0.9
# ---
# datatype:
# - {name: c, datatype: float64, format: .2f, description: description-c}
# - {name: tm.jd1, datatype: float64}
# - {name: tm.jd2, datatype: float64}
# - {name: sc.x, unit: '', datatype: float64}
# - {name: sc.y, unit: '', datatype: float64}
# - {name: sc.z, unit: '', datatype: float64}
# - {name: q, unit: m, datatype: float64, format: .3f, description: description-q}
# - {name: lat, unit: deg, datatype: float64, description: description-lat}
# - {name: ang, unit: deg, datatype: float64, description: description-ang}
# - {name: el.x, unit: m, datatype: float64, description: description-el}
# - {name: el.y, unit: m, datatype: float64, description: description-el}
# - {name: el.z, unit: m, datatype: float64, description: description-el}
# meta: !!omap
# - __mixin_columns__:
#     ang:
#       __class__: astropy.coordinates.angles.Angle
#       __info__: {description: description-ang}
#       unit: &id001 !astropy.units.Unit {unit: deg}
#       value: !astropy.table.SerializedColumn {name: ang}
#     el:
#       __class__: astropy.coordinates.earth.EarthLocation
#       __info__: {description: description-el}
#       ellipsoid: WGS84
#       x: !astropy.table.SerializedColumn {name: el.x}
#       y: !astropy.table.SerializedColumn {name: el.y}
#       z: !astropy.table.SerializedColumn {name: el.z}
#     lat:
#       __class__: astropy.coordinates.angles.Latitude
#       __info__: {description: description-lat}
#       unit: *id001
#       value: !astropy.table.SerializedColumn {name: lat}
#     q:
#       __class__: astropy.units.quantity.Quantity
#       __info__: {description: description-q, format: .3f}
#       unit: !astropy.units.Unit {unit: m}
#       value: !astropy.table.SerializedColumn {name: q}
#     sc:
#       __class__: astropy.coordinates.sky_coordinate.SkyCoord
#       __info__: {description: description-sc}
#       frame: icrs
#       representation: cartesian
#       x: !astropy.table.SerializedColumn {name: sc.x}
#       y: !astropy.table.SerializedColumn {name: sc.y}
#       z: !astropy.table.SerializedColumn {name: sc.z}
#     tm:
#       __class__: astropy.time.core.Time
#       __info__: {description: description-tm}
#       format: cxcsec
#       in_subfmt: '*'
#       jd1: !astropy.table.SerializedColumn {name: tm.jd1}
#       jd2: !astropy.table.SerializedColumn {name: tm.jd2}
#       out_subfmt: '*'
#       precision: 5
#       scale: tt
```

mhvk commented on astropy/io/ascii/ecsv.py:
```diff
@@ -210,3 +227,43 @@ class Ecsv(basic.Basic):
+        table : `astropy.table.Table`
+            Output table for writing
+        """
+        @contextlib.contextmanager
```
A quite horrible hack, but I learned something new -- this `contextmanager` is quite neat...

taldcroft replied:
On the dog walk I thought of a better way -- a new TimeInfo attribute `serialize_as` which is a dict that indicates how various serialization engines should behave.  So something like:
```
class TimeInfo(...):
     serialize_formatted_values = {'fits': False, 'ecsv': True, 'hdf5': False, 'yaml': False}
```
In theory this could be put into `MixinInfo`.  Not sure if that should be done before there is a use case, but probably `SkyCoord` will want this.  For `SkyCoord` this would choose between the (new) formatted output vs. the decimal in the default preferred units.

mhvk commented on astropy/table/serialize.py:
```diff
@@ -0,0 +1,165 @@
+                # Column so the ECSV header looks complete (and for back compatibility
+                # with pre-2.0 behavior).  However, these are ignored in object
+                # construction later in favor of values in the
+                # __mixin_columns__['__info__'] dict.  This is because in general
```
True, but as shown in my PR to your branch, one can of course choose to get the info from the single column in the case that there is only one serialized column. I think that is still nicer than to duplicate the information.

mhvk commented on astropy/table/serialize.py:
```diff
@@ -0,0 +1,165 @@
+                    col_info = {}
+                else:
+                    col_info = info
+                new_cols.append(Column(data, name=name, **col_info))  # TODO masking, MaskedColumn
```
Ah, so much nicer than what I had!! So, I would suggest setting `info = {}` for the single-col case.

mhvk commented on astropy/table/serialize.py:
```diff
@@ -0,0 +1,165 @@
+                for attr in col.info.attrs_from_parent:
+                    if attr in info:
+                        del info[attr]
+                obj_attrs['__info__'] = info
```
Probably should check here again if there still is anything left in `info`.

mhvk commented on astropy/table/serialize.py:
```diff
@@ -0,0 +1,165 @@
+    mod_name, cls_name = re.match(r'(.+)\.(\w+)', cls_full_name).groups()
+    module = import_module(mod_name)
+    cls = getattr(module, cls_name)
+    return cls.info._construct_from_dict(obj_attrs)
```
I guess one solution to the constructor not setting `info` attributes would be to change this `_construct_from_dict` to also take `info_attrs`. It might be as simple as making a new method `MixinInfo._set_info_attributes()`? Though just doing it explicitly below is fine for now too.

mhvk commented on astropy/table/serialize.py:
```diff
@@ -0,0 +1,165 @@
+            obj_attrs[data_attr] = out[name]
+            del out[name]
+
+        info = obj_attrs.pop('__info__', {})
```
So, here one would again want to distinguish between the single-column and the multi-column case, getting `info` from the serialized column in the single-column case (but excluding empty things).

taldcroft commented:
I'm thinking about point 1 "I think the split between data_attrs and info_attrs will, in the end, turn out not to be necessary (since some info will need to be serialized in columns), so I would suggest to keep them together".  I have thought about this a lot through this project and appreciate your questioning how this is done.

The crux of the matter is here in serialize.py:
```
    for col in tbl.itercols():
        if not has_info_class(col, MixinInfo):
            new_cols.append(col)
        else:
            obj_attrs = col.info._represent_as_dict()
            data_attrs = col.info._represent_as_dict_data_attrs
```
Here is the point when the representer needs to be explicitly informed which of the attributes to break out as columns.  If it is just given a dict of stuff it simply cannot know.  We could have it guess (pick everything that looks like it might be well-suited for a column), but I want to let the parent column class explicitly be responsible for how things get serialized.  That is the job of `_represent_as_dict_data_attrs`, which as you can see is now a property in `Time`.  It should also be made a property in `SkyCoord` but I just didn't get there.

If you think naming it `_data_attrs` is better then I can do that, but I sort of think the tight association with `represent_as_dict` is useful to indicate.  In any case this really is all internal and we can just change it any time as long as astropy tests are passing.

taldcroft commented:
> Should we, for now, explicitly fail on non-scalar info attributes? (Probably not, but asking in case.)

Why would you want that?  Certainly `meta` is a non-scalar info attribute that must be allowed.  Or something like `location` can be non-scalar, so it just gets stored as YAML.  We can make things work better in a back compatible way as needed in the future.  I do think that very large arrays for `location` or `obstime` are going to be a fairly rare corner case.

taldcroft commented:
> Should SerializedColumn also store the __class__, so that for instance for jd1 and jd2, we don't have to hack Time to ensure columns get turned into proper arrays?

This was really a bug in `Time` which I have fixed in a better way now.  This is astropy 1.3:
```
In [2]: c = Column([1,2], format='.2f', name='WHAT THE HECK??????')
In [3]: tm = Time(c, format='jd')
In [4]: tm
Out[4]: 
<Time object: scale='utc' format='jd' value=WHAT THE HECK??????
-------------------
               1.00
               2.00>
```
In general, where appropriate, astropy classes should accept `Column` objects as initializers without affecting how the object behaves downstream.

taldcroft commented:
Given that the release freeze is in a day and I still need to get the Time masking PR finished up...  is this in a place where the external API (code and ECSV output) is good enough?   I think that future improvements could be made back-compatible (new code can read old files).  So if/when we make some clean-up improvements it won't be a big deal.

taldcroft commented:
And of course notice the new `info.serialize_method` attribute which will be relevant for FITS output. 

cc: @AustereCuriosity

mhvk commented:
@taldcroft - see https://github.com/taldcroft/astropy/pull/14 for recursive serialization. I also ensured that if there is just a single column, information is not duplicated (no reason to treat `dtype` different from `unit`, I think).

I think this looks quite good. Smaller todo's (bit too tired to fix those now):
- Ensure `Longitude.wrap_angle` gets stored (needs its own `info` I suppose).
- Get `NDArrayMixin` to work, so that `Table` can store `EarthLocation` as well.

Note that with my PR, there is no need for separate `data` and `info` attributes any more (but I did not change that back, as I want to be sure you like the scheme first).

mhvk commented:
@taldcroft - I added 2 further commits to https://github.com/taldcroft/astropy/pull/14 - to ensure `Longitude.wrap_angle` is preserved, and that angle subclasses work well with table operations(I think doing `empty_like` for them would be a problem, since those "empty" values can easily be out of range for `Latitude`; so I fill with zeros instead).

mhvk commented:
Left-overs:
- With my PR, we can remove the separation between `data_attrs` and `info_attrs` - it is all decided based on shape. Shall I go ahead and revert that change?
- I'm still a bit torn about `EarthLocation`; shouldn't we just make it a mixin for `Table` as well? It is more chance than design that it is based on `Quantity` -- it makes just as much (if not more) sense as a representation (indeed, see #4261). Maybe the test for in `_add_as_mixin_column` can be changed to `has_info_class(col, MixinInfo) and not has_info_class(col, QuantityInfo)`?

taldcroft commented:
Go ahead on both items.  If you can do the later in a clean way without any regressions or unrelated test changes then I'll good with the concept.

mhvk commented:
OK, I did both - see two further commits to https://github.com/taldcroft/astropy/pull/14

More or less the end of the day here, but I think it has really come together! So, I would suggest merging my commits, and then perhaps having a last look -- though I think this will be all ready to go in.

mhvk commented:
@taldcroft - on the issue of mixin columns, I cannot say I have a good solution... One idea might be to have a table class `TableBase` which does not change the nature of any column, and then the final conversion would call `Table` or `QTable` on that, converting whatever is needed. Alternatively, maybe one should just call the final conversion unconditionally.  (A propos that line, wouldn't we want a `copy_data=False` there?).

taldcroft commented:
@mhvk - see 1a78461 for what I think is a reasonable solution:

- First, it within `serialize.py` it constructs the table as a `QTable`.  This means we are no longer relying on a bug of `add_column` and can fix that at some point.
- Then the ECSV ascii.read() will downgrade to `Table` if-and-only-if there are no Quantity mixins.  This is the key and seems reasonable to me.  It is consistent with previous behavior (since the ECSV file itself could not previously store mixins directly).  In other words any old ECSV file will be returned as `Table` by `ascii.read`.  But a new (`astropy-2.0` schema) ECSV with Quantity mixins will be read as QTable by default.  And a new ECSV without Quantity mixins will also be a `Table`.
- The `Table.read` and `QTable.read` methods always return the right class, and now bypass some class coercion within `io.registry.read` which does a copy.

taldcroft commented:
I really like your idea of making `EarthLocationInfo` not be a subclass of `QuantityInfo`.  This allows a more nuanced way of distinguishing between classes that are accidental subclasses but need to be treated differently for whatever reason.

taldcroft commented:
I think that this is now (finally) feature-complete!

mhvk commented on astropy/table/table.py:
```diff
@@ -2521,7 +2521,22 @@ def read(cls, *args, **kwargs):
+        # Always read as Table because all astropy readers initially return
+        # Table or QTable.  Most readers return Table but e.g. ascii.ecsv might
+        # return QTable.
+        out = io_registry.read(Table, *args, **kwargs)
```
I understand the post-read conversion, but not quite why the input to registry read should be changed -- won't many readers do it correctly already?  Indeed, if I undo this locally, tests still pass.

mhvk commented on docs/io/ascii/write.rst:
```diff
@@ -269,3 +269,95 @@ details.
+The `Enhanced Character-Separated Values (ECSV) format
+<https://github.com/astropy/astropy-APEs/blob/master/APE6.rst>`_ can be used to
+write astropy `~astropy.table.Table` or `~astropy.table.QTable` datasets to a
+text-only data file and then read the table back with loss of information.  The
```
Surely should be "with*out*" loss of information!

mhvk replied:
Looking at this again - did you mean that there might be slight loss of precision? If so, just update...

taldcroft replied:
It was supposed to be *without*.

mhvk commented on docs/io/ascii/write.rst:
```diff
@@ -269,3 +269,95 @@ details.
+  >>> t = QTable([c, q, sc], names=['c', 'q', 'sc'])
+
+  >>> ascii.write(t, format='ecsv')   # doctest: +SKIP
+  # %ECSV 0.9
```
this needs updating - I'll make a quick PR.

mhvk commented:
@taldcroft - I pushed the fix and documentation updates directly to your branch, since I thought they were trivial enough. With this, I think this is ready to go in!

And I agree that this was a very nice example of us working together towards what is not only is a neat way to represent mixin data, but also one for which the output is quite readable. And, on top of that, it has set us up very well to do the same for other formats!

taldcroft commented:
Whew!  Thanks so much for making this a much better thing in the end.

Speaking of applications, I was thinking this morning that we may way to use this for generating the `repr` or `str` of tables with mixin columns.  It might be pretty trivial and fix up the current awkwardness of representing multi-component columns like SkyCoord or EarthLocation.

mhvk commented:
Yes, since this is such a nice format even just in `ecsv`, it does make sense to do something similar for just `repr`. Especially if we can get `SkyCoord` to be a bit more like `Time`, and have a nice `format` associated...
